Add Alibaba Cloud AnalyticDB Spark Support

dcoliversun opened the pull request:
This PR aims to add Alibaba Cloud AnalyticDB Spark[1] support.
1. implement custom operator/hook/sensor for AnalyticDB Spark.
2. enable users to schedule AnalyticDB Spark Batch/SQL applications by DAGs.
3. AnalyticDB Spark hooks provides log/Web UI url/application status for users.

[1]: https://www.alibabacloud.com/help/en/analyticdb-for-mysql/latest/spark-offline-application-development
closes: #31738

utkarsharma2 commented on airflow/providers/alibaba/cloud/hooks/analyticdb_spark.py:
```diff
@@ -0,0 +1,377 @@
+            return self.get_adb_spark_client().submit_spark_app(request)
+        except Exception as e:
+            self.log.error(e)
+            raise AirflowException("Errors when submit spark application")
```
```suggestion
            self.log.error(e)
            raise AirflowException("Errors when submit spark application") from e
```

utkarsharma2 replied:
Nice to chain exception here.

utkarsharma2 commented on airflow/providers/alibaba/cloud/hooks/analyticdb_spark.py:
```diff
@@ -0,0 +1,377 @@
+            return self.get_adb_spark_client().submit_spark_app(request)
+        except Exception as e:
+            self.log.error(e)
+            raise AirflowException("Errors when submit spark sql")
```
```suggestion
            raise AirflowException("Errors when submit spark sql") from e
```

utkarsharma2 commented on airflow/providers/alibaba/cloud/hooks/analyticdb_spark.py:
```diff
@@ -0,0 +1,377 @@
+            ).body.data.state
+        except Exception as e:
+            self.log.error(e)
+            raise AirflowException(f"Errors when fetching state for spark application: {app_id}")
```
```suggestion
            raise AirflowException(f"Errors when fetching state for spark application: {app_id}") from e
```

utkarsharma2 commented on airflow/providers/alibaba/cloud/hooks/analyticdb_spark.py:
```diff
@@ -0,0 +1,377 @@
+            ).body.data.web_ui_address
+        except Exception as e:
+            self.log.error(e)
+            raise AirflowException(f"Errors when fetching web ui address for spark application: {app_id}")
```
```suggestion
            raise AirflowException(f"Errors when fetching web ui address for spark application: {app_id}") from e
```

utkarsharma2 commented on airflow/providers/alibaba/cloud/hooks/analyticdb_spark.py:
```diff
@@ -0,0 +1,377 @@
+            ).body.data.log_content
+        except Exception as e:
+            self.log.error(e)
+            raise AirflowException(f"Errors when fetching log for spark application: {app_id}")
```
```suggestion
            raise AirflowException(f"Errors when fetching log for spark application: {app_id}") from e
```

utkarsharma2 commented on airflow/providers/alibaba/cloud/hooks/analyticdb_spark.py:
```diff
@@ -0,0 +1,377 @@
+            self.get_adb_spark_client().kill_spark_app(KillSparkAppRequest(app_id=app_id))
+        except Exception as e:
+            self.log.error(e)
+            raise AirflowException(f"Errors when killing spark application: {app_id}")
```
```suggestion
            raise AirflowException(f"Errors when killing spark application: {app_id}") from e
```

utkarsharma2 commented on airflow/providers/alibaba/cloud/hooks/analyticdb_spark.py:
```diff
@@ -0,0 +1,377 @@
+        extra_config = self.adb_spark_conn.extra_dejson
+        auth_type = extra_config.get("auth_type", None)
+        if not auth_type:
+            raise Exception("No auth_type specified in extra_config.")
```
Maybe a more specific exception type will be nice, so that code handling it doesn't have to use `except Exception` which can potentially mask all other errors.

utkarsharma2 replied:
I think replacing every instance of `raise Exception` where we know the cause of the issue would be nice.

dcoliversun replied:
Agreed with you.   
Replace Exception with ValueError. I summarize the scenarios where Exception is used:
1. Spark application parameters specified by user;
2. Connection content;

utkarsharma2 commented on airflow/providers/alibaba/cloud/hooks/analyticdb_spark.py:
```diff
@@ -0,0 +1,377 @@
+                data=self.build_submit_sql_data(*args, **kwargs),
+                app_type="SQL"
+            )
+            return self.get_adb_spark_client().submit_spark_app(request)
```
```suggestion
            request = SubmitSparkAppRequest(
                dbcluster_id=cluster_id,
                resource_group_name=rg_name,
                data=self.build_submit_sql_data(*args, **kwargs),
                app_type="SQL"
            )
        try:
            return self.get_adb_spark_client().submit_spark_app(request)
```

utkarsharma2 replied:
I'm assuming you only expect `self.get_adb_spark_client().submit_spark_app(request)` to raise an exception and not request the creation part?

phanikumv commented on airflow/providers/alibaba/cloud/hooks/analyticdb_spark.py:
```diff
@@ -0,0 +1,377 @@
+
+        :param app_id: identifier of the spark application
+
+        :return: application state
```
```suggestion
```

phanikumv commented on airflow/providers/alibaba/cloud/hooks/analyticdb_spark.py:
```diff
@@ -0,0 +1,377 @@
+
+        :param app_id: identifier of the spark application
+
+        :return: web ui address for application
```
```suggestion
```

sunank200 commented on airflow/providers/alibaba/cloud/hooks/analyticdb_spark.py:
```diff
@@ -0,0 +1,371 @@
+class AppState(Enum):
+    """AnalyticDB Spark application states"""
+
+    SUBMITTED = "SUBMITTED"
```
Is there a documentation link for these app states? We can add that in doc-strings

dcoliversun replied:
Added :)

sunank200 commented on airflow/providers/alibaba/cloud/sensors/analyticdb_spark.py:
```diff
@@ -0,0 +1,72 @@
+        self._adb_spark_conn_id = adb_spark_conn_id
+        self._adb_spark_hook: AnalyticDBSparkHook | None = None
+
+    def get_hook(self) -> AnalyticDBSparkHook:
```
this can be `cached_property`. More at: https://docs.python.org/3/library/functools.html

dcoliversun replied:
Thanks :)

sunank200 reviewed (commented):
Thanks for your contribution

eladkal commented on airflow/providers/alibaba/CHANGELOG.rst:
```diff
@@ -24,6 +24,14 @@
 Changelog
 ---------
 
+2.5.0
```
Please remove this entry.
Change log is auto generates by release manager

eladkal commented on airflow/providers/alibaba/provider.yaml:
```diff
@@ -23,6 +23,7 @@ description: |
 
 suspended: false
 versions:
+  - 2.5.0
```
Please remove version. It is set by release manager

dcoliversun committed:
Add Alibaba Cloud AnalyticDB Spark Support

Apply suggestions from code review

Co-authored-by: Utkarsh Sharma <utkarsharma2@gmail.com>

replace Exception with ValueError

fix ci deps failure

add version

remove return in comment

fix indent error

address comments

address comments

eladkal commented on airflow/utils/db.py:
```diff
@@ -477,6 +477,19 @@ def create_default_connections(session: Session = NEW_SESSION):
+    merge_conn(
+        Connection(
+            conn_id="adb_spark_default",
+            conn_type="adb_spark",
```
Where is `adb` coming from? (I understand it's for AnalyticDB but is it something you came up with?
I didn't find `adb` to officially used in Alibaba cloud docs.

On another front, cc also @ephraimbuddy @pierrejeambrun would you prefer to have this change in a separated PR? This file change is a core change that needs to be released with Airflow core not with providers. This also means that the connection will be available only to Airflow 2.7.0+
WDYT?

dcoliversun replied:
> Where is `adb` coming from?

`adb` is short name for `analyticdb` officially and doc is https://www.alibabacloud.com/help/en/analyticdb-for-mysql. short name usage is like https://www.alibabacloud.com/help/en/analyticdb-for-mysql/latest/sdk-reference

> would you prefer to have this change in a separated PR?

I'm ok and I will revert `db.py` in this PR. User can use other connection like this content in adb spark.

ephraimbuddy replied:
@eladkal , Is the provider going to be released before 2.7.0? I think it shouldn't since it touched core because the connection entry is needed to use it in any Airflow version, right?.

eladkal replied:
I created https://github.com/apache/airflow/issues/32048 to hopefully solve it once and for all

As for your question - the provider must support minimum version of Airflow 2.4.0

dcoliversun committed:
revert db.py change

phanikumv commented on airflow/providers/alibaba/cloud/hooks/analyticdb_spark.py:
```diff
@@ -0,0 +1,371 @@
+        Check the values in the provided list can be converted to strings.
+
+        :param vals: list to validate
+        :return: true if valid
```
```suggestion
```

dcoliversun replied:
All `:return:` removed

phanikumv commented on airflow/providers/alibaba/cloud/hooks/analyticdb_spark.py:
```diff
@@ -0,0 +1,371 @@
+        Check configuration values are either strings or ints.
+
+        :param conf: configuration variable
+        :return: true if valid
```
```suggestion
```

phanikumv commented on airflow/providers/alibaba/cloud/operators/analyticdb_spark.py:
```diff
@@ -0,0 +1,228 @@
+        """
+        Get valid hook.
+
+        :return: hook
```
```suggestion
```

phanikumv commented on tests/providers/alibaba/cloud/hooks/test_analyticdb_spark.py:
```diff
@@ -0,0 +1,193 @@
+            self.hook = AnalyticDBSparkHook(adb_spark_conn_id=MOCK_ADB_SPARK_CONN_ID)
+
+    def test_build_submit_app_data(self):
+        res_data = self.hook.build_submit_app_data(
```
docstring missing, please describe what you are trying to test here

dcoliversun replied:
All test add docstring

phanikumv commented on tests/providers/alibaba/cloud/hooks/test_analyticdb_spark.py:
```diff
@@ -0,0 +1,193 @@
+        assert res_data == except_data
+
+    def test_build_submit_sql_data(self):
+        res_data = self.hook.build_submit_sql_data(
```
docstring missing, please describe what you are trying to test here

phanikumv commented on tests/providers/alibaba/cloud/hooks/test_analyticdb_spark.py:
```diff
@@ -0,0 +1,193 @@
+
+    @mock.patch(ADB_SPARK_STRING.format("AnalyticDBSparkHook.get_adb_spark_client"))
+    def test_submit_spark_app(self, mock_service):
+        # Given
```
docstring missing, please describe what you are trying to test here

phanikumv commented on tests/providers/alibaba/cloud/hooks/test_analyticdb_spark.py:
```diff
@@ -0,0 +1,193 @@
+
+    @mock.patch(ADB_SPARK_STRING.format("AnalyticDBSparkHook.get_adb_spark_client"))
+    def test_submit_spark_sql(self, mock_service):
+        # Given
```
docstring missing, please describe what you are trying to test here

phanikumv commented on tests/providers/alibaba/cloud/hooks/test_analyticdb_spark.py:
```diff
@@ -0,0 +1,193 @@
+
+    @mock.patch(ADB_SPARK_STRING.format("AnalyticDBSparkHook.get_adb_spark_client"))
+    def test_get_spark_state(self, mock_service):
+        # Given
```
docstring missing, please describe what you are trying to test here

phanikumv commented on tests/providers/alibaba/cloud/hooks/test_analyticdb_spark.py:
```diff
@@ -0,0 +1,193 @@
+
+    @mock.patch(ADB_SPARK_STRING.format("AnalyticDBSparkHook.get_adb_spark_client"))
+    def test_get_spark_web_ui_address(self, mock_service):
+        # Given
```
docstring missing, please describe what you are trying to test here

phanikumv commented on tests/providers/alibaba/cloud/hooks/test_analyticdb_spark.py:
```diff
@@ -0,0 +1,193 @@
+
+    @mock.patch(ADB_SPARK_STRING.format("AnalyticDBSparkHook.get_adb_spark_client"))
+    def test_get_spark_log(self, mock_service):
+        # Given
```
docstring missing, please describe what you are trying to test here

dcoliversun committed:
address comments

dcoliversun committed:
address comments

dcoliversun committed:
Merge branch 'main' into AIRFLOW-31738

dcoliversun committed:
fix ut error

dcoliversun committed:
fix static check

dcoliversun committed:
fix static check

eladkal commented on docs/apache-airflow-providers-alibaba/index.rst:
```diff
@@ -83,13 +83,15 @@ for the minimum Airflow version supported) via
+============================  ==================
+PIP package                   Version required
+============================  ==================
+``apache-airflow``            ``>=2.5.0``
```
This is wrong.
The version in provider.yaml is 
`apache-airflow>=2.4.0`

you need to rerun the pre-commits so this file would be generated correctly

dcoliversun committed:
fix index.rst
